eureka/S3_data_reduction/wfc3.py

```diff
@@ -0,0 +1,498 @@
+        The metadata object
+    m:  int
+        The current file number
+
+    '''
+
+    # Write drift2D
+    def writeDrift2D(arg):
+        drift2D, m, n = arg
+        # Assign to array of spectra and uncertainties
+        meta.drift2D[-1][n] = drift2D
+        return
+    
+    # Save the reference frame for each scan direction if not yet done
+    if m<2:
+        # FINDME: This requires that the reference files be the first two files.
+        # Using other files as the reference files will require loading in all of the frames at once
+        # This will still work for observations with only one scan direction, since the second ref file will never be used
+        meta.subdata_ref.append(data.subdata)
+        meta.diffmask_ref.append(meta.diffmask[-1])
+
+    print("Calculating 2D drift...")
+    #FINDME: instead of calculating scanHeight, consider fitting stretch factor
+    drift2D  = np.zeros((meta.nreads-1, 2))
+    meta.drift2D.append(drift2D)
+    if meta.ncpu == 1:
+        # Only 1 CPU
+        # Get index of reference frame
+        # (0 = forward scan, 1 = reverse scan)
+        p   = meta.scandir[m]
+        for n in range(meta.nreads-1):
+            writeDrift2D(hst.calcDrift2D(meta.subdata_ref[p][0]*meta.subdiffmask[p][0],
+                                         data.subdata[n]*meta.subdiffmask[-1][n], m, n, meta.num_data_files))
+    else:
+        # Multiple CPUs
+        pool = mp.Pool(meta.ncpu)
+        # Get index of reference frame
+        # (0 = forward scan, 1 = reverse scan)
+        p   = meta.scandir[m]
+        for n in range(meta.nreads-1):
+            res = pool.apply_async(hst.calcDrift2D, args=(meta.subdata_ref[p][0]*meta.subdiffmask[p][0],
+                                                          data.subdata[n]*meta.subdiffmask[-1][n],
+                                                          m, n, meta.num_data_files), callback=writeDrift2D)
+        pool.close()
+        pool.join()
+        res.wait()
+    
+    print("Performing rough, pixel-scale drift correction...")
+    meta.drift2D_int.append(np.round(meta.drift2D[-1],0))
+    # Correct for drift by integer pixel numbers, no interpolation
+    for n in range(meta.nreads-1):
+        data.subdata[n]     = spni.shift(data.subdata[n],  -1*meta.drift2D_int[-1][n,::-1], order=0,
+                                         mode='constant', cval=0)
+        data.submask[n]     = spni.shift(data.submask[n],  -1*meta.drift2D_int[-1][n,::-1], order=0,
+                                         mode='constant', cval=0)
+        data.subvariance[n] = spni.shift(data.subvariance[n], -1*meta.drift2D_int[-1][n,::-1], order=0,
+                                         mode='constant', cval=0)
+        data.subbg[n]       = spni.shift(data.subbg[n],    -1*meta.drift2D_int[-1][n,::-1], order=0,
+                                         mode='constant', cval=0)
+
+    # FINDME: The following cannot be run since we don't have the full time axis
```

Comment:
This could be a problem for some datasets if we can't run this step.

Reply:
Yeah, I'm still thinking of how we can best enable this option without writing a completely different S3 for WFC3. Removing this one step was one of the biggest concessions I had to make with this first attempt and is something I plan to continue to work on. I can try to do so for this PR if you want, but it'll likely take me some time to work out a solution

Reply:
I ran into this same kind of issue with SPCA earlier. It is cetainly doable to rewrite everything so that we can sigma clip along the full time-axis, but it ends up being very RAM intensive since you need to store at least 2 copies of all the frames in RAM to do sigma clipping. Plenty doable with just a transit/eclipse, but it gets rough with a phase curve (and even worse if it's on JWST's larger detectors).

Reply:
Let's hold off on making changes inside this PR. A good compromise is to load in chunks of data to avoid ram issues. We could break things up by HST orbit or by a specific number of frames.